File: typescipt/existing/src/serviceProviderExtensions.ts







namespace Lexy.Compiler;

public static class ServiceProviderExtensions
{
   public static IServiceCollection AddLexy(this IServiceCollection services)
   {
     services.TryAdd(ServiceDescriptor.Scoped<ILexyParser, LexyParser>());
     services.TryAdd(ServiceDescriptor.Scoped<IParserLogger, ParserLogger>());
     services.TryAdd(ServiceDescriptor.Scoped<IParserContext, ParserContext>());
     services.TryAdd(ServiceDescriptor.Scoped<ISourceCodeDocument, SourceCodeDocument>());
     services.TryAdd(ServiceDescriptor.Scoped<ITokenizer, Tokenizer>());

     services.TryAdd(ServiceDescriptor.Scoped<IExecutionEnvironment, ExecutionEnvironment>());
     services.TryAdd(ServiceDescriptor.Scoped<IExecutionContext, ExecutionContext>());

     services.TryAdd(ServiceDescriptor.Scoped<ILexyCompiler, LexyCompiler>());

     services.TryAdd(ServiceDescriptor.Scoped<ISpecificationRunnerContext, SpecificationRunnerContext>());
     services.TryAdd(ServiceDescriptor.Scoped<ISpecificationsRunner, SpecificationsRunner>());
     services.TryAdd(ServiceDescriptor.Scoped<ISpecificationFileRunner, SpecificationFileRunner>());
     services.TryAdd(ServiceDescriptor.Scoped<IScenarioRunner, ScenarioRunner>());
     services.TryAdd(ServiceDescriptor.Transient<ISpecificationsRunner, SpecificationsRunner>());

     return services;
   }
}

File: src/Lexy.Compiler/ServiceProviderExtensions.cs
using System;
using Lexy.Compiler.Compiler;
using Lexy.Compiler.FunctionLibraries;
using Lexy.Compiler.Infrastructure;
using Lexy.Compiler.Language.Expressions;
using Lexy.Compiler.Parser;
using Lexy.Compiler.Parser.Tokens;
using Lexy.Compiler.Specifications;
using Lexy.RunTime;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.DependencyInjection.Extensions;

namespace Lexy.Compiler;

public static class ServiceProviderExtensions
{
    public static IServiceCollection AddLexy(this IServiceCollection services, params Type[] libraries)
    {
        return services.Singleton<ILexyParser, LexyParser>()

            .Singleton<ISourceCodeDocument, SourceCodeDocument>()
            .Singleton<ITokenizer, Tokenizer>()
            .Singleton<IExpressionFactory, ExpressionFactory>()

            .Singleton<IFileSystem, FileSystem>()
            .Singleton<IExecutionContext, ExecutionContext>()
            .SingletonFactory<ILibraries, Libraries>(_ => new Libraries(libraries))

            .Singleton<ILexyCompiler, LexyCompiler>()

            .AddScoped<ICompilationEnvironment, CompilationEnvironment>()
            .Transient<ISpecificationsRunner, SpecificationsRunner>();
    }

    private static IServiceCollection Singleton<TInterface, IImplementation>(this IServiceCollection services)
        where TInterface : class
        where IImplementation : class, TInterface
    {
        services.TryAdd(ServiceDescriptor.Singleton<TInterface, IImplementation>());

        return services;
    }

    private static IServiceCollection SingletonFactory<TInterface, IImplementation>(this IServiceCollection services,
        Func<IServiceProvider,TInterface> factory)
        where TInterface : class
        where IImplementation : class, TInterface
    {
        services.Replace(ServiceDescriptor.Singleton(factory));

        return services;
    }

    private  static IServiceCollection Transient<TInterface, IImplementation>(this IServiceCollection services)
        where TInterface : class
        where IImplementation : class, TInterface
    {
        services.TryAdd(ServiceDescriptor.Transient<TInterface, IImplementation>());

        return services;
    }
}